# FIT5196 Task 2 in Assessment 1
#### Student Name: Boyu Zhang
#### Student ID: 28491300

Date: XX/XX/XXXX

Version: 1.0

Environment: Python 3.6 and Jupyter notebook

Libraries used: please include the main libraries you used in your assignment here, e.g.,:
```
* re (for regular expression, included in Anaconda Python 3.6) 
* pandas (to save the cleaned result into the output which is a csv file)
* pdfminer.six (to extract the content of the table and save them in a txt)
```

## 1.  Import libraries 

In [1]:
# Code to import libraries as you need in this assessment, e.g.,
import re
import csv
import pdfminer
import pandas as pd

## 2. Parse PDF File

In this section, you can write your python scripts to parse the correspondiing file.
You should 
* write proper notes for all code block in this notebook using the Markdown cells
* provide proper comment in your scripts
* run all cells to make sure scripts are runable. If the scripts cannot be run by the assessors, they will not be assessed and zero mark will be given to the task.

### 2.1 Pre-processing -- parse the pdf to a txt file
The first step is to extract the table from the pdf file and save them to a txt file by running the following command in the terminal:

`pdf2txt.py -o healthpdf.txt ./Input/health.pdf`

This extract all the data(including the needed and the not needed ones) and save them in a txt file nameed healthpdf.txt 

**Note**: The txt can be simply generated by doing the above command in the prompt which can be quite different on different operating system, my system is linux so the command may not be compatitble with you tutor's. Thus I have included the txt file in the ZIP package to make life easier. I can guarantee the txt is just a copy of the 
original pdf in a messy way with no manual manipulation.

In [2]:
#the path of the txt file which will be the input in the following process
pdfTextFile = './healthpdf.txt'
with open(pdfTextFile,'r') as f:
    lines = f.readlines()
#have a exploration of the txt:

for line in lines:
    print(repr(line))#the repr function return of string of printable  representation of the content

'TABLE 3. HEALTH\n'
'\n'
'Use of basic \n'
'drinking water \n'
'services (%)\n'
'\n'
'Use of basic \n'
'sanitation \n'
'services (%)\n'
'\n'
'2015\n'
'\n'
'2015\n'
'\n'
'BCG DTP1β DTP3β polio3 MCV1 MCV2^ HepB3 Hib3 rota PCV3\n'
'\n'
'total urban rural\n'
'\n'
'total urban rural\n'
'\n'
'2016\n'
'\n'
'2011–2016*\n'
'\n'
'Immunization coverage (%)\n'
'\n'
'Pneumonia Diarrhoea\n'
'\n'
'Malaria\n'
'\n'
'Protection \n'
'\n'
'at birth \n'
'(PAB) \n'
'against \n'
'tetanusλ \n'
'\n'
'Care seeking \n'
'for children \n'
'\n'
'(%)\n'
'\n'
'with  \n'
'\n'
'symptoms of \n'
'pneumonia \n'
'\n'
'Treatment \n'
'with oral \n'
'rehydra-\n'
'tion salts \n'
'(ORS) (%)\n'
'\n'
'Care \n'
'\n'
'seeking for \n'
'children \n'
'with fever \n'
'\n'
'(%)\n'
'\n'
'Countries \n'
'and areas\n'
'63\n'
'Afghanistan\n'
'91\n'
'Albania\n'
'93\n'
'Algeria\n'
'100\n'
'Andorra\n'
'41\n'
'Angola\n'
'98\n'
'Anguilla\n'
'97\n'
'Antigua and Barbuda\n'
'100\n'
'Argentina\n'
'99\n'
'Armenia\n'
'100\n'
'Australia\n'
'100\n'
'Aust

'–\n'
'66\n'
'98\n'
'–\n'
'24\n'
'100\n'
'99\n'
'99\n'
'76\n'
'28\n'
'76\n'
'28\n'
'100\n'
'89\n'
'95\n'
'58\n'
'88\n'
'58\n'
'99\n'
'100\n'
'96\n'
'94\n'
'94\n'
'97\n'
'73\n'
'28\n'
'–\n'
'97\n'
'–\n'
'98\n'
'99\n'
'94\n'
'–\n'
'92\n'
'\n'
'100\n'
'100\n'
'63\n'
'6\n'
'27\n'
'–\n'
'98\n'
'99\n'
'48\n'
'100\n'
'59\n'
'13\n'
'81\n'
'58\n'
'72\n'
'98\n'
'100\n'
'–\n'
'–\n'
'70\n'
'68\n'
'76\n'
'64\n'
'–\n'
'92\n'
'–\n'
'96\n'
'–\n'
'28\n'
'–\n'
'35\n'
'91\n'
'–\n'
'8\n'
'–\n'
'98\n'
'99\n'
'18\n'
'8\n'
'69\n'
'6\n'
'100\n'
'95\n'
'99\n'
'23\n'
'61\n'
'58\n'
'100\n'
'100\n'
'89\n'
'96\n'
'96\n'
'83\n'
'30\n'
'5\n'
'93\n'
'92\n'
'–\n'
'83\n'
'89\n'
'99\n'
'–\n'
'91\n'
'\n'
'–\n'
'–\n'
'98\n'
'77\n'
'64\n'
'99\n'
'–\n'
'99\n'
'85\n'
'–\n'
'99\n'
'89\n'
'99\n'
'90\n'
'76\n'
'94\n'
'32\n'
'97\n'
'97\n'
'97\n'
'84\n'
'96\n'
'99\n'
'96\n'
'96\n'
'99\n'
'76\n'
'–\n'
'92\n'
'98\n'
'97\n'
'98\n'
'99\n'
'92\n'
'99\n'
'0\n'
'–\n'
'98\n'
'37\n'
'74\n'
'37\n'
'–\n'
'99\n'
'99\n'
'96\n'
'–\n'
'97\n'
'2

### 2.2 Colllecting all the country names
From the above result we can see most country names share a comman pattern such as 
```
'Afghanistan\n'
```
However, we should not ignore more special cases like:
```
'Bolivia (Plurinational State of)\n'
'Central African Republic \n'
```
There are even some errors in the country names where two columns are merged together/
the long name are seperated in two strings, the crossing line is easily found in the cases where the country name is too long:
```
'Democratic Republic of the Congo 42\n'

'Democratic People's Republic of \n'
'Korea\n'

'The former Yugoslav Republic of \n'
'Macedonia'

```

#### Remove the table headers of all four pages
#### where a part of the headers is not in the correct place,thus need to be corrected

In [3]:
# collect the index of all table headers and unnecessary area of the tables from all 4 pages
positions = []
for index, line in enumerate(lines):
    if "TABLE" in line:
        header = []
        header.append(index)
    if line.startswith('and areas'):
        header.append(index)
        positions.append(header)
    if line.startswith('SUMMARY'):
        positions.append(index)
    

In [4]:
positions

[[0, 62], [1523, 1526], [3070, 3073], [4618, 4621], 5108]

In [5]:
lines[:] = lines[positions[0][1]  : positions[1][0]] +\
           lines[positions[1][1]  : positions[2][0]] +\
           lines[positions[2][1]  : positions[3][0]] +\
           lines[positions[3][1]  : positions[-1]]


In [6]:
#collect the index of mis-placed table headers in the following 3 pages and drop them
mis_header = []
for index, line in enumerate(lines):
    if line.startswith('Im'):
        p = []
        p.append(index)
    if line.startswith('2011'):
        p.append(index)
        mis_header.append(p)

In [7]:
mis_header

[[1526, 1583], [3071, 3128], [4566, 4635]]

In [8]:
lines[:] = lines[:mis_header[0][0]] +\
           lines[mis_header[0][1] + 1: mis_header[1][0] ] +\
           lines[mis_header[1][1] + 1: mis_header[2][0] ] +\
           lines[mis_header[2][1] + 1: ]

In [9]:
for line in lines:
     print(repr(line))

'and areas\n'
'63\n'
'Afghanistan\n'
'91\n'
'Albania\n'
'93\n'
'Algeria\n'
'100\n'
'Andorra\n'
'41\n'
'Angola\n'
'98\n'
'Anguilla\n'
'97\n'
'Antigua and Barbuda\n'
'100\n'
'Argentina\n'
'99\n'
'Armenia\n'
'100\n'
'Australia\n'
'100\n'
'Austria\n'
'84\n'
'Azerbaijan\n'
'98\n'
'Bahamas\n'
'100\n'
'Bahrain\n'
'97\n'
'Bangladesh\n'
'98\n'
'Barbados\n'
'98\n'
'Belarus\n'
'100\n'
'Belgium\n'
'97\n'
'Belize\n'
'67\n'
'Benin\n'
'98\n'
'Bhutan\n'
'93\n'
'Bolivia (Plurinational State of)\n'
'98\n'
'Bosnia and Herzegovina\n'
'79\n'
'Botswana\n'
'97\n'
'Brazil\n'
'100\n'
'British Virgin Islands\n'
'100\n'
'Brunei Darussalam\n'
'99\n'
'Bulgaria\n'
'54\n'
'Burkina Faso\n'
'56\n'
'Burundi\n'
'86\n'
'Cabo Verde\n'
'75\n'
'Cambodia\n'
'65\n'
'Cameroon\n'
'99\n'
'Canada\n'
'54\n'
'Central African Republic \n'
'43\n'
'Chad\n'
'100\n'
'Chile\n'
'96\n'
'China\n'
'97\n'
'Colombia\n'
'84\n'
'Comoros\n'
'68\n'
'Congo\n'
'100\n'
'Cook Islands\n'
'100\n'
'Costa Rica\n'
'73\n'
"Côte d'Ivoire\n"
'100\n'
'Croatia\

'97\n'
'98\n'
'99\n'
'92\n'
'99\n'
'0\n'
'–\n'
'98\n'
'37\n'
'74\n'
'37\n'
'–\n'
'99\n'
'99\n'
'96\n'
'–\n'
'97\n'
'26\n'
'–\n'
'66\n'
'98\n'
'99\n'
'99\n'
'85\n'
'79\n'
'–\n'
'80\n'
'–\n'
'95\n'
'96\n'
'98\n'
'–\n'
'98\n'
'\n'
'98\n'
'92\n'
'99\n'
'87\n'
'64\n'
'99\n'
'99\n'
'99\n'
'79\n'
'99\n'
'96\n'
'87\n'
'99\n'
'94\n'
'87\n'
'99\n'
'99\n'
'99\n'
'98\n'
'94\n'
'96\n'
'97\n'
'99\n'
'99\n'
'95\n'
'99\n'
'95\n'
'78\n'
'97\n'
'98\n'
'96\n'
'97\n'
'98\n'
'97\n'
'98\n'
'99\n'
'98\n'
'99\n'
'52\n'
'78\n'
'35\n'
'99\n'
'99\n'
'99\n'
'97\n'
'92\n'
'96\n'
'99\n'
'99\n'
'61\n'
'98\n'
'99\n'
'97\n'
'95\n'
'93\n'
'–\n'
'83\n'
'97\n'
'99\n'
'99\n'
'99\n'
'–\n'
'99\n'
'\n'
'95\n'
'92\n'
'98\n'
'67\n'
'49\n'
'99\n'
'96\n'
'99\n'
'72\n'
'98\n'
'73\n'
'72\n'
'93\n'
'89\n'
'86\n'
'98\n'
'98\n'
'98\n'
'98\n'
'89\n'
'89\n'
'97\n'
'98\n'
'97\n'
'95\n'
'98\n'
'62\n'
'66\n'
'96\n'
'98\n'
'93\n'
'92\n'
'96\n'
'84\n'
'97\n'
'96\n'
'94\n'
'99\n'
'42\n'
'66\n'
'26\n'
'97\n'
'99\n'
'99\n'
'93\n'
'91\n'
'90\n'

In [10]:
#remaining mis_headers
mis_header2 = []
for index, line in enumerate(lines):
    if line.startswith('Child'):
        p = []
        p.append(index)
    if line.startswith('ITN (%)'):
        p.append(index)
        mis_header2.append(p)


In [11]:
mis_header2

[[1323, 1333], [2807, 2817], [4294, 4304]]

In [12]:
countriesData = lines[:mis_header2[0][0]] +\
            lines[mis_header2[0][1] + 1 : mis_header2[1][0]] +\
            lines[mis_header2[1][1] + 1: mis_header2[2][0]] +\
            lines[mis_header2[2][1] + 1 :]

In [13]:
for line in countriesData:
    print(repr(line))

'and areas\n'
'63\n'
'Afghanistan\n'
'91\n'
'Albania\n'
'93\n'
'Algeria\n'
'100\n'
'Andorra\n'
'41\n'
'Angola\n'
'98\n'
'Anguilla\n'
'97\n'
'Antigua and Barbuda\n'
'100\n'
'Argentina\n'
'99\n'
'Armenia\n'
'100\n'
'Australia\n'
'100\n'
'Austria\n'
'84\n'
'Azerbaijan\n'
'98\n'
'Bahamas\n'
'100\n'
'Bahrain\n'
'97\n'
'Bangladesh\n'
'98\n'
'Barbados\n'
'98\n'
'Belarus\n'
'100\n'
'Belgium\n'
'97\n'
'Belize\n'
'67\n'
'Benin\n'
'98\n'
'Bhutan\n'
'93\n'
'Bolivia (Plurinational State of)\n'
'98\n'
'Bosnia and Herzegovina\n'
'79\n'
'Botswana\n'
'97\n'
'Brazil\n'
'100\n'
'British Virgin Islands\n'
'100\n'
'Brunei Darussalam\n'
'99\n'
'Bulgaria\n'
'54\n'
'Burkina Faso\n'
'56\n'
'Burundi\n'
'86\n'
'Cabo Verde\n'
'75\n'
'Cambodia\n'
'65\n'
'Cameroon\n'
'99\n'
'Canada\n'
'54\n'
'Central African Republic \n'
'43\n'
'Chad\n'
'100\n'
'Chile\n'
'96\n'
'China\n'
'97\n'
'Colombia\n'
'84\n'
'Comoros\n'
'68\n'
'Congo\n'
'100\n'
'Cook Islands\n'
'100\n'
'Costa Rica\n'
'73\n'
"Côte d'Ivoire\n"
'100\n'
'Croatia\

'100\n'
'100\n'
'97\n'
'94\n'
'99\n'
'94\n'
'94\n'
'99\n'
'\n'
'100\n'
'100\n'
'97\n'
'89\n'
'82\n'
'–\n'
'100\n'
'95\n'
'92\n'
'100\n'
'99\n'
'84\n'
'99\n'
'95\n'
'96\n'
'99\n'
'100\n'
'–\n'
'–\n'
'96\n'
'100\n'
'99\n'
'77\n'
'–\n'
'98\n'
'–\n'
'99\n'
'–\n'
'83\n'
'–\n'
'91\n'
'88\n'
'–\n'
'75\n'
'100\n'
'99\n'
'100\n'
'90\n'
'70\n'
'97\n'
'60\n'
'100\n'
'96\n'
'86\n'
'73\n'
'98\n'
'95\n'
'100\n'
'100\n'
'99\n'
'92\n'
'99\n'
'96\n'
'91\n'
'90\n'
'–\n'
'100\n'
'–\n'
'100\n'
'99\n'
'91\n'
'–\n'
'100\n'
'\n'
'100\n'
'100\n'
'61\n'
'36\n'
'54\n'
'–\n'
'100\n'
'78\n'
'87\n'
'97\n'
'87\n'
'29\n'
'98\n'
'72\n'
'86\n'
'96\n'
'100\n'
'–\n'
'–\n'
'79\n'
'100\n'
'90\n'
'49\n'
'–\n'
'98\n'
'–\n'
'95\n'
'–\n'
'74\n'
'–\n'
'63\n'
'95\n'
'–\n'
'47\n'
'–\n'
'97\n'
'99\n'
'56\n'
'20\n'
'63\n'
'48\n'
'100\n'
'91\n'
'94\n'
'52\n'
'88\n'
'60\n'
'100\n'
'100\n'
'94\n'
'68\n'
'97\n'
'98\n'
'60\n'
'45\n'
'100\n'
'100\n'
'–\n'
'83\n'
'100\n'
'98\n'
'–\n'
'99\n'
'\n'
'98\n'
'100\n'
'76\n'
'13\n'
'33\n'
'97\n'

In [14]:
#design a regular expression pattern to extract all the country names
p1 = r'^([A-Z])(.+?)\s*\d*\n'

In [15]:
finalCountries = []
for c in countriesData:
    m = re.match(p1,c)
    if m:
        finalCountries.append(m.group(1) + m.group(2))
        

#### Fix the country names that cross two lines

In [16]:
position = []
for index, c in enumerate(finalCountries):
    if c == 'Democratic People\'s Republic of':
        position.append(index)

position

[48]

In [17]:
finalCountries[position[0] : position[0] + 2] = [finalCountries[position[0]] + " " +finalCountries[position[0] + 1]]

In [18]:
position = []
for index, c in enumerate(finalCountries):
    if c == 'The former Yugoslav Republic of':
        position.append(index)
position

[177]

In [19]:
finalCountries[position[0] : position[0] + 2] = [finalCountries[position[0]] + " " +finalCountries[position[0] + 1]]

In [20]:
#check the number of countries is correct
len(finalCountries)

202

### 2.3 Extract all columns

#### first fix the the only data of the first column that is falsely merged into the country name

In [21]:
for index, line in enumerate(countriesData):
        if 'Democratic Republic of the Congo' in line:
            p = index
p2 = '^([A-z].+?)\s*(\d*)\n'
match = re.match(p2,countriesData[p])
countriesData = countriesData[:p] + [match.group(2),match.group(1)] + countriesData[p + 1:]

In [22]:
for index,line in enumerate(countriesData):
    print(index,repr(line))

0 'and areas\n'
1 '63\n'
2 'Afghanistan\n'
3 '91\n'
4 'Albania\n'
5 '93\n'
6 'Algeria\n'
7 '100\n'
8 'Andorra\n'
9 '41\n'
10 'Angola\n'
11 '98\n'
12 'Anguilla\n'
13 '97\n'
14 'Antigua and Barbuda\n'
15 '100\n'
16 'Argentina\n'
17 '99\n'
18 'Armenia\n'
19 '100\n'
20 'Australia\n'
21 '100\n'
22 'Austria\n'
23 '84\n'
24 'Azerbaijan\n'
25 '98\n'
26 'Bahamas\n'
27 '100\n'
28 'Bahrain\n'
29 '97\n'
30 'Bangladesh\n'
31 '98\n'
32 'Barbados\n'
33 '98\n'
34 'Belarus\n'
35 '100\n'
36 'Belgium\n'
37 '97\n'
38 'Belize\n'
39 '67\n'
40 'Benin\n'
41 '98\n'
42 'Bhutan\n'
43 '93\n'
44 'Bolivia (Plurinational State of)\n'
45 '98\n'
46 'Bosnia and Herzegovina\n'
47 '79\n'
48 'Botswana\n'
49 '97\n'
50 'Brazil\n'
51 '100\n'
52 'British Virgin Islands\n'
53 '100\n'
54 'Brunei Darussalam\n'
55 '99\n'
56 'Bulgaria\n'
57 '54\n'
58 'Burkina Faso\n'
59 '56\n'
60 'Burundi\n'
61 '86\n'
62 'Cabo Verde\n'
63 '75\n'
64 'Cambodia\n'
65 '65\n'
66 'Cameroon\n'
67 '99\n'
68 'Canada\n'
69 '54\n'
70 'Central African Republi

1088 '–\n'
1089 '–\n'
1090 '91\n'
1091 '85\n'
1092 '89\n'
1093 '87\n'
1094 '–\n'
1095 '92\n'
1096 '93\n'
1097 '–\n'
1098 '95\n'
1099 '–\n'
1100 '92\n'
1101 '85\n'
1102 '92\n'
1103 '93\n'
1104 '85\n'
1105 '–\n'
1106 '60\n'
1107 '80\n'
1108 '–\n'
1109 '–\n'
1110 '90\n'
1111 '85\n'
1112 '85\n'
1113 '–\n'
1114 '–\n'
1115 '90\n'
1116 '–\n'
1117 '–\n'
1118 '–\n'
1119 '–\n'
1120 '97\n'
1121 '85\n'
1122 '–\n'
1123 '85\n'
1124 '–\n'
1125 '90\n'
1126 '88\n'
1127 '80\n'
1128 '90\n'
1129 '70\n'
1130 '94\n'
1131 '–\n'
1132 '80\n'
1133 '94\n'
1134 '\n'
1135 '62\n'
1136 '70 x\n'
1137 '66\n'
1138 '–\n'
1139 '49\n'
1140 '–\n'
1141 '–\n'
1142 '94\n'
1143 '57 x\n'
1144 '–\n'
1145 '–\n'
1146 '36 x\n'
1147 '–\n'
1148 '–\n'
1149 '42\n'
1150 '–\n'
1151 '93\n'
1152 '–\n'
1153 '67\n'
1154 '23\n'
1155 '74 x\n'
1156 '62\n'
1157 '87\n'
1158 '14 x\n'
1159 '50 x\n'
1160 '–\n'
1161 '–\n'
1162 '–\n'
1163 '52\n'
1164 '63\n'
1165 '–\n'
1166 '69\n'
1167 '28\n'
1168 '–\n'
1169 '30 x\n'
1170 '26\n'
1171 '–\n'
1172 '–\n'
1

2837 '–\n'
2838 '–\n'
2839 '–\n'
2840 '73\n'
2841 '43\n'
2842 '–\n'
2843 '–\n'
2844 '79\n'
2845 '–\n'
2846 '–\n'
2847 '18\n'
2848 '–\n'
2849 '–\n'
2850 '–\n'
2851 '–\n'
2852 '–\n'
2853 '–\n'
2854 '–\n'
2855 '–\n'
2856 '36\n'
2857 '19\n'
2858 '6\n'
2859 '–\n'
2860 '–\n'
2861 '\n'
2862 '–\n'
2863 '–\n'
2864 '36\n'
2865 '69\n'
2866 '–\n'
2867 '–\n'
2868 '68\n'
2869 '–\n'
2870 '–\n'
2871 '–\n'
2872 '47\n'
2873 '90\n'
2874 '5\n'
2875 '19\n'
2876 '–\n'
2877 '–\n'
2878 '–\n'
2879 '–\n'
2880 '–\n'
2881 '3 x\n'
2882 '–\n'
2883 '–\n'
2884 '–\n'
2885 '–\n'
2886 '–\n'
2887 '–\n'
2888 '–\n'
2889 '–\n'
2890 '–\n'
2891 '59\n'
2892 '–\n'
2893 '–\n'
2894 '–\n'
2895 '50\n'
2896 '–\n'
2897 '–\n'
2898 '–\n'
2899 '62\n'
2900 '–\n'
2901 '–\n'
2902 '–\n'
2903 '–\n'
2904 '80\n'
2905 '57\n'
2906 '–\n'
2907 '–\n'
2908 '93\n'
2909 '–\n'
2910 '–\n'
2911 '67\n'
2912 '–\n'
2913 '–\n'
2914 '–\n'
2915 '–\n'
2916 '–\n'
2917 '–\n'
2918 '–\n'
2919 '–\n'
2920 '51\n'
2921 '27\n'
2922 '24\n'
2923 '–\n'
2924 '–\n'
2925 '\n'

4587 '89\n'
4588 '71\n'
4589 '0\n'
4590 '92\n'
4591 '99\n'
4592 '0\n'
4593 '53\n'
4594 '95\n'
4595 '49\n'
4596 '58\n'
4597 '63\n'
4598 '\n'
4599 '78\n'
4600 '26\n'
4601 '99\n'
4602 '0\n'
4603 '97\n'
4604 '93\n'
4605 '95\n'
4606 '99\n'
4607 '64\n'
4608 '84\n'
4609 '96\n'
4610 '71\n'
4611 '91\n'
4612 '90\n'
4613 '\n'
4614 '78\n'
4615 '47\n'
4616 '99\n'
4617 '94\n'
4618 '97\n'
4619 '93\n'
4620 '95\n'
4621 '99\n'
4622 '64\n'
4623 '84\n'
4624 '96\n'
4625 '71\n'
4626 '91\n'
4627 '90\n'
4628 '\n'
4629 '0\n'
4630 '0\n'
4631 '92\n'
4632 '90\n'
4633 '96\n'
4634 '73\n'
4635 '0\n'
4636 '99\n'
4637 '0\n'
4638 '47\n'
4639 '0\n'
4640 '59\n'
4641 '90\n'
4642 '91\n'
4643 '\n'
4644 '78\n'
4645 '0\n'
4646 '99\n'
4647 '92\n'
4648 '96\n'
4649 '93\n'
4650 '94\n'
4651 '99\n'
4652 '0\n'
4653 '7\n'
4654 '0\n'
4655 '71\n'
4656 '90\n'
4657 '90\n'
4658 '\n'
4659 '87\n'
4660 '–\n'
4661 '–\n'
4662 '–\n'
4663 '90\n'
4664 '–\n'
4665 '–\n'
4666 '–\n'
4667 '78\n'
4668 '75\n'
4669 '94\n'
4670 '70\n'
4671 '85\n'
4672 '80

In [23]:
#check the number of values that can be matched
count = 0
p3 = '\d{1,3}|–'
for line in countriesData:
    if re.match(p3,line):
        count += 1
count

4509

#### With 202 countries and 22 colmun we should have `4444` recors not `4509` which means there are columns that we don't need but we haven't dropped them
**It appears that there are 5 columns of the last page is placed incorrectly which will certainly cause problem if we take cells merely by counting the number of records.**

**Need to be fixed!!!!**

In [24]:
#remove the unneeded columns which belong to the summary but is exactly read and palced before "SUMMARY":
countriesData[:] = countriesData[:4689] + countriesData[4702:4718] + countriesData[4731:4747] +\
countriesData[4760:4776] + countriesData[4789:4805] + countriesData[4818:]

In [25]:
#check the number of values again:
count = 0
for line in countriesData:
    if re.match(p3,line):
        count += 1
count

4444

then we need to find out the number records in every page:

In [26]:
pageBounds = {}
for index, c in enumerate(finalCountries):
    if c in ['Fiji','Nepal','Tuvalu','Zimbabwe']:
        pageBounds[c] = index 
pageBounds


{'Fiji': 61, 'Nepal': 124, 'Tuvalu': 187, 'Zimbabwe': 201}

this means each page has rescpectively **62,63,63,14** records

In [27]:
recodrsPerpage =[62,63,63,14]


#### Now we can start to capture data for all 22 columns

In [28]:
num_of_cols = 22
data = {}
for i in range(num_of_cols):
    data[i] = []


In [29]:
pageNum = -1 
numRecs = 0
colIdx = 0
idx = 0
for line in countriesData:
    if line.startswith('and areas'):
        pageNum += 1
        numRecs = recodrsPerpage[pageNum]
        colIdx = 0
        idx = 0
    if re.match(p3,line) and colIdx < num_of_cols:
        data[colIdx].append(re.match(p3,line).group(0))
        idx += 1
        if idx % numRecs == 0:
            colIdx += 1

### 2.4 Store data in csv format

In [30]:
#load the data into a pandas datafram
df = pd.DataFrame(data,columns=list(range(22)))
#add the column of country names
df.insert(0,column='Country Name',value=finalCountries)
#replace the dash with empty string
df = df.apply(lambda x: x.str.replace('–',''),axis=1);df

Country Name    0    1    2    3    4    5  \
0                                  Afghanistan   63   89   53   39   56   33   
1                                      Albania   91   93   90   98   98   97   
2                                      Algeria   93   95   89   87   90   82   
3                                      Andorra  100  100  100  100  100  100   
4                                       Angola   41   63   23   39   62   21   
5                                     Anguilla   98   98        97   97        
6                          Antigua and Barbuda   97             88             
7                                    Argentina  100  100  100   95   95   94   
8                                      Armenia   99   99   99   92   96   83   
9                                    Australia  100  100  100  100             
10                                     Austria  100  100  100  100  100  100   
11                                  Azerbaijan   84   95   72   89   92   87   
12                                     Bahamas   98             92             
13                                     Bahrain  100            100             
14                                  Bangladesh   97   98   97   47   54   43   
15                                    Barbados   98             96             
16                                     Belarus   98   98   99   94   94   95   
17                                     Belgium  100  100  100   99   99   99   
18                                      Belize   97   99   96   87   91   84   
19                                       Benin   67   77   60   14   25    5   
20                                      Bhutan   98   97   98   63   72   57   
21            Bolivia (Plurinational State of)   93   99   79   53   64   27   
22                      Bosnia and Herzegovina   98   97   98   95   99   92   
23                                    Botswana   79   95   58   60   75   39   
24                                      Brazil   97   99   87   86   91   58   
25                      British Virgin Islands  100             97             
26                           Brunei Darussalam  100  100   99   96   96   97   
27                                    Bulgaria   99   99   99   86   87   84   
28                                Burkina Faso   54   79   43   23   48   12   
29                                     Burundi   56   88   52   50   46   51   
..                                         ...  ...  ...  ...  ...  ...  ...   
172                                     Sweden  100  100  100   99   99  100   
173                                Switzerland  100  100  100  100  100  100   
174                       Syrian Arab Republic   97   99   94   93   96   89   
175                                 Tajikistan   74   92   68   95   94   96   
176                                   Thailand   98   99   97   95   94   96   
177  The former Yugoslav Republic of Macedonia   97   96   98   91   97   83   
178                                Timor-Leste   70   91   60   44   73   30   
179                                       Togo   63   90   45   14   28    5   
180                                    Tokelau  100       100   93        93   
181                                      Tonga  100  100  100   93   97   92   
182                        Trinidad and Tobago   97             92             
183                                    Tunisia   94  100   83   93   98   83   
184                                     Turkey   99   99  100   96   99   89   
185                               Turkmenistan   94   91   98   97   94   99   
186                   Turks and Caicos Islands   94             88             
187                                     Tuvalu   99  100   99   91   92   91   
188                                     Uganda   39   73   32   19   28   17   
189                                    Ukraine   98   97  100   96   97   93   
190                       United A

In [31]:
df.to_csv('./Output/health.csv',index=False)

## 3. Summary
Give a short summary of your work done above, such as your findings.
>So in this task I have suffered a lot in parsing the pdf to the sample output.
Due the excessively complex header structure of table, the uneeded information are spreaded all over
the place which took me a lot of time to fix them.
The most meaningful lesson is that pdf should definitely be avoided for storing large dataset!

## 4.Reference
- Merge some list items in a Python List.Retrieved from https://stackoverflow.com/questions/1142851/merge-some-list-items-in-a-python-list
- `pandas` Dataframe. Retrieved from https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html